### PR TITLE
[Bug Fix] Fix `TrotterProduct` to be compatible with resource tracking and error tracking

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
`qml.TrotterProduct()` now has error tracking functionality, but ins't compatible with resource tracking. This bug fix adds native resource tracking functionality to allow for compatibility with both pipelines.

image

**Description of the Change:**
Inherit from `ResourcesOperation` class as well and directly implement the resource method. 

**Benefits:**
Allows for tracking resources without decomposing the template. 

**Possible Drawbacks:**
If we change the target gate-set or want to decompose further, then this method is no longer accurate.

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/templates/subroutines/trotter.py
Comment (soranjh) [inline pennylane/templates/subroutines/trotter.py:1]: Could you add the failing example as a test?

File: pennylane/templates/subroutines/trotter.py [hunk 1]
```diff
@@ -236,6 +239,30 @@ def queue(self, context=qml.QueuingManager):
...
+        gate_types = defaultdict(int)
+        gate_sizes = defaultdict(int)
+
+        for op in decomp:
+            gate_types[op.name] += 1
+            gate_sizes[len(op.wires)] += 1
+
+        return Resources(num_wires, num_gates, gate_types, gate_sizes, depth)
+
     def error(
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/templates/subroutines/trotter.py:266]: I changed both to have the "-bound" suffix, for consistency.

File: pennylane/templates/subroutines/trotter.py [hunk 2]
```diff
@@ -144,13 +147,13 @@ def my_circ():
 
         An *upper-bound* for the error in approximating time-evolution using this operator can be
         computed by calling :func:`~.TrotterProduct.error()`. It is computed using two different methods; the
-        "one-norm" scaling method and the "commutator" scaling method. (see `Childs et al. (2021) <https://arxiv.org/abs/1912.08854>`_)
+        "one-norm-bound" scaling method and the "commutator-bound" scaling method. (see `Childs et al. (2021) <https://arxiv.org/abs/1912.08854>`_)
```
Comment (soranjh) [inline pennylane/templates/subroutines/trotter.py:150]: Probably need to update the changelog in this PR.

File: tests/templates/test_subroutines/test_trotter.py [hunk 3]
```diff
@@ -211,6 +213,93 @@
     }
 )
 
+test_resources_data = {
+    (0, 1): Resources(
```
Comment (soranjh) [inline tests/templates/test_subroutines/test_trotter.py:217]: A comment here defining `(0, 1)` will be nice.

File: tests/templates/test_subroutines/test_trotter.py [hunk 4]
```diff
@@ -211,6 +213,93 @@
     }
 )
 
+test_resources_data = {
```
Comment (soranjh) [inline tests/templates/test_subroutines/test_trotter.py:216]: You can bring it close to the class that uses it.

File: tests/templates/test_subroutines/test_trotter.py [hunk 5]
```diff
@@ -565,6 +654,74 @@ def test_tensorflow_interface(self):
             _ = op.error()
 
 
+class TestResources:
+    """Test the resources methof of the TrotterProduct class"""
```
Comment (soranjh) [inline tests/templates/test_subroutines/test_trotter.py:658]:
```suggestion
    """Test the resources method of the TrotterProduct class"""
```

**Review by soranjh — COMMENTED** [review]
Thanks @Jaybsoni, looks good to go.

File: tests/templates/test_subroutines/test_trotter.py (hunk 3 above)
Comment (Jaybsoni) [reply tests/templates/test_subroutines/test_trotter.py:217]: Its consistent with the other tests for trotter (see for example the `TestDecomposition`), I can add a comment

File: tests/templates/test_subroutines/test_trotter.py (hunk 4 above)
Comment (Jaybsoni) [reply tests/templates/test_subroutines/test_trotter.py:216]: The other data tuples are also defined at the top, so I leave it here for consistency.

File: doc/releases/changelog-0.36.0.md [hunk 6]
```diff
@@ -681,6 +681,9 @@
 * `qml.equal` can now be used with sums and products that contain operators on no wires like `I` and `GlobalPhase`.
   [(#5562)](https://github.com/PennyLaneAI/pennylane/pull/5562)
 
+* Fixed `qml.TrotterProduct` to be compatible with resource tracking by inheirting from `ResourcesOperation`.
+  [(#5629)](https://github.com/PennyLaneAI/pennylane/pull/5629)
```
Comment (Jaybsoni) [inline doc/releases/changelog-0.36.0.md:685]:
```suggestion
* Fixed `qml.TrotterProduct` to be compatible with resource tracking by inheriting from `ResourcesOperation`. Updated the names for the `method` kwarg of the `error` method from "one-norm" and "commutator" to "one-norm-bound" and "commutator-bound".
  [(#5629)](https://github.com/PennyLaneAI/pennylane/pull/5629)
```

File: pennylane/templates/subroutines/trotter.py (hunk 2 above)
Comment (Jaybsoni) [reply pennylane/templates/subroutines/trotter.py:150]: Updated

File: tests/templates/test_subroutines/test_trotter.py (hunk 3 above)
Comment (Jaybsoni) [inline tests/templates/test_subroutines/test_trotter.py:217]:
```suggestion
    (0, 1): Resources(  # (hamiltonian_index, order): Resources computed by hand
```

Comment (Jaybsoni) [reply tests/templates/test_subroutines/test_trotter.py]: Added

**Review by soranjh — CHANGES_REQUESTED** [review]
Thanks @Jaybsoni, please split the PR. Thanks.

File: tests/templates/test_subroutines/test_trotter.py [hunk 7]
```diff
@@ -211,6 +213,93 @@
     }
 )
 
+test_resources_data = {
+    (0, 1): Resources(  # (hamiltonian_index, order): Resources computed by hand
```
Comment (obliviateandsurrender) [inline tests/templates/test_subroutines/test_trotter.py:217]:
```suggestion
    # (hamiltonian_index, order): Resources computed by hand
    (0, 1): Resources( 
```

**Review by obliviateandsurrender — CHANGES_REQUESTED** [review]
Thanks @Jaybsoni! Leaving a query and couple of suggestions.

File: pennylane/templates/subroutines/trotter.py [hunk 8]
```diff
@@ -236,8 +239,32 @@ def queue(self, context=qml.QueuingManager):
...
+        Returns:
+            Resources: The resources for an instance of TrotterProduct.
+        """
+        with qml.QueuingManager.stop_recording():
+            decomp = self.compute_decomposition(*self.parameters, **self.hyperparameters)
+
+        num_wires = len(self.wires)
+        num_gates = len(decomp)
+
+        unique_operation_decomp = (copy.deepcopy(op) for op in decomp)
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/templates/subroutines/trotter.py:254]: The problem with the existing decomposition of TrotterProduct, is that it reuses the same operators (with qml.apply). The `unique_operation_decomp` generator is required to make a new copy of each operator in the decomposition.

**Comment by astralcai** [comment]
As discussed in our standup meeting today, this PR is not expected to be included in the v0.36 release. Can we rebase the PR to the master branch? @Jaybsoni @trbromley

**Comment by trbromley** [comment]
@Jaybsoni what is easier here, opening a new PR or rebasing this one? CC @DSGuala

**Pull request closed without merging**